### PR TITLE
Updated digest tests with new local digest server

### PR DESCRIPTION
**Pull request by atulbi**
- Added new digest auth server and updated existing digest auth tests.
- Added test for the wrong password
- Added skipped test with existing authorization header

#### Why we need a test for existing authorization header
Currently, we do not remove `Authorization` header in the pre-hook. Users might provide `Authorization` header with the request.
Auth server expects unsigned request to have no headers on it. If headers are not removed in the pre-hook, server might throw 400 status code.

#### Solution
For removing the headers in pre-hook we need to have `request`, but `request` is not accessible in the pre-hook right now.
So, added a skipped test in which server returns 400 error code if `Authorization` header is provided.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,12 +33,15 @@
     "aws4": "1.9.1",
     "crypto-js": "3.3.0",
     "eventemitter3": "4.0.0",
+    "express": "^4.17.1",
```
Comment (codenirvana) [inline package.json:36]: Add as `devDependencies` and use the absolute version, not `^`.

File: test/fixtures/server.js [hunk 2]
```diff
@@ -609,6 +612,36 @@ function createBytesServer () {
     return server;
 }
 
+/**
+ * creates an Digest server
+ *
+ * @param {Object} options - The options for the server
+ * @param {String} options.username - Username for authentication
+ * @param {String} options.password - Password for authentication
+ *
```
Comment (codenirvana) [inline test/fixtures/server.js:621]: Extra newline

File: test/fixtures/server.js [hunk 3]
```diff
@@ -609,6 +612,36 @@ function createBytesServer () {
     return server;
 }
 
+/**
+ * creates an Digest server
```
Comment (codenirvana) [inline test/fixtures/server.js:616]: capitalization

File: test/fixtures/server.js [hunk 4]
```diff
@@ -609,6 +612,36 @@ function createBytesServer () {
...
+        expectedUsername = options.username || 'username',
+        expectedPassword = options.password || 'password';
+
+    passport.use(new DigestStrategy({qop: 'auth'},
+        (username, done) => {
+            return done(null, username === expectedUsername ? username : false, expectedPassword);
+        }
+    ));
+
+    app.get('/private',
```
Comment (codenirvana) [inline test/fixtures/server.js:635]: expose routes similar to NTLM and edgeGrid servers.

File: test/fixtures/server.js [hunk 5]
```diff
@@ -609,6 +612,36 @@ function createBytesServer () {
...
+        passport.authenticate('digest', {session: false}),
+        (req, res) => {
+            res.send(req.users);
+        }
+    );
+
+    return app.listen(3000);
```
Comment (codenirvana) [inline test/fixtures/server.js:642]: don't hardcode port.

File: test/integration/auth-methods/digest.test.js [hunk 6]
```diff
@@ -594,4 +595,48 @@ describe('digest auth', function () {
             expect(authHeader).to.match(/qop=auth/);
         });
     });
+
+    // This test fails because authorization header is present during first request.Depending upon
+    // server implementation, server might return 400 - Bad Request.
+    // Any existing Authorization header must be removed in the pre-hook for Digest Auth.
+    it.skip('should fail when Authorization header is present', function (done) {
```
Comment (codenirvana) [inline test/integration/auth-methods/digest.test.js:602]: Replace all instances of `https://postman-echo.com/digest-auth`.

File: test/integration/auth-methods/digest.test.js [hunk 7]
```diff
@@ -1,7 +1,25 @@
-var expect = require('chai').expect;
+var expect = require('chai').expect,
+    {createDigestServer} = require('../../fixtures/server');
 
 describe('digest auth', function () {
-    var testrun;
+    var PORT = 2000,
```
Comment (codenirvana) [inline test/integration/auth-methods/digest.test.js:5]: don't hardcode port number

File: test/integration/auth-methods/digest.test.js [hunk 8]
```diff
@@ -1,7 +1,25 @@
-var expect = require('chai').expect;
+var expect = require('chai').expect,
+    {createDigestServer} = require('../../fixtures/server');
```
Comment (codenirvana) [inline test/integration/auth-methods/digest.test.js:2]: ES6 🚨

File: test/integration/auth-methods/digest.test.js [hunk 9]
```diff
@@ -1,7 +1,25 @@
...
-    var testrun;
+    var PORT = 2000,
+        USERNAME = 'postman',
+        PASSWORD = 'password',
+        digestServerURL = 'http://localhost:' + PORT,
+        digestServer,
+        testrun;
+
+    before(function () {
+        digestServer = createDigestServer({
```
Comment (codenirvana) [inline test/integration/auth-methods/digest.test.js:13]: `createDigestServer` should not start the server.

File: test/integration/auth-methods/digest.test.js [hunk 10]
```diff
@@ -1,7 +1,25 @@
...
+            password: PASSWORD,
+            PORT
+        });
+    });
+
+    after(function () {
+        digestServer.destroy();
+    });
 
     // @todo
```
Comment (codenirvana) [inline test/integration/auth-methods/digest.test.js:24]: check if possible now

Comment (atulbi) [reply test/integration/auth-methods/digest.test.js:29]: still not possible, as this server also return qop-value

**Pull request merged**